[imagedecoder] change image decoder addon system to new way

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->
This change the addon system headers from "C" to "C++", further is this the
first who use the new way to get extensions from addon.xml. In this the
mimetype and supported file extensions.

This extension way allow to check several parts on different types without
load of addon.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/ImageDecoder.h:
```diff
@@ -0,0 +1,163 @@
+    kodi::addon::CInstanceImageDecoder* addonInstance;
+    bool (__cdecl* load_image_from_memory) (AddonInstance_ImageDecoder const* instance,
+                                         unsigned char* buffer, unsigned int buf_size,
+                                         unsigned int* width, unsigned int* height);
```
indentation?

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/ImageDecoder.h:
```diff
@@ -0,0 +1,163 @@
+      m_instanceData->toAddon.decode = ADDON_Decode;
+    }
+
+    inline static bool ADDON_LoadImageFromMemory(AddonInstance_ImageDecoder const* instance,
```
"type const *" is the same as "const type *". Why do you use other syntax here than you do usually? Is your intention to denote that the pointer value itself is const and the data behind the pointer can be written? Then you must use "type * const". Applies also to other methods in the PR.

ksooo commented on xbmc/guilib/imagefactory.cpp:
```diff
@@ -46,16 +46,15 @@ IImage* ImageFactory::CreateLoader(const CURL& url)
-  VECADDONS codecs;
-  ADDON::CBinaryAddonCache &addonCache = CServiceBroker::GetBinaryAddonCache();
-  addonCache.GetAddons(codecs, ADDON::ADDON_IMAGEDECODER);
-  for (auto& codec : codecs)
```
Does this mean that we will no longer need binary addon cache once all your changes are in? Is <code>CServiceBroker::GetBinaryAddonManager().GetAddonInfos()</code> performance-wise comparable with  <code>CServiceBroker::GetBinaryAddonCache().GetAddons()</code>?

AlwinEsch replied:
Yes, this replace complete the binary addon cache, mostly is only the info like here for supported types needed.

This parts are continuous stored in binary addon manager.

In case now a running addon need to be asked becomes them during active time also stored and can be used by new way.

From performance it is equal with the other.

AlwinEsch replied:
Have measured performance (with one addon present).

|  |    |
|--|---|
| With binary addon cache | ~0.000040 sec
| Complete without cache | ~0.005000 sec
| With new on this | ~0.000065 sec

ksooo replied:
Okay, thanks for clarifying.

ksooo reviewed (commented):
some notes/questions.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/ImageDecoder.h:
```diff
@@ -0,0 +1,163 @@
+    kodi::addon::CInstanceImageDecoder* addonInstance;
+    bool (__cdecl* load_image_from_memory) (const AddonInstance_ImageDecoder* instance,
+                                         unsigned char* buffer, unsigned int buf_size,
+                                         unsigned int* width, unsigned int* height);
```
indentation?

ksooo reviewed (approved):
Only a small indentation glitch, otherwise good to go.

AlwinEsch committed:
[imagedecoder] change image decoder addon system to new way

This change the addon system headers from "C" to "C++", further is this the
first who use the new way to get extensions from addon.xml. In this the
mimetype and supported file extensions.

This extension way allow to check several parts on different types without
load of addon.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
@notspiff OK from your side to bring in?

notspiff commented:
sure
